### PR TITLE
Add shared memory buffer video driver

### PR DESCRIPTION
**Pull request by visigoth**
These changes add a video driver that sources its data from an external process via a shared memory buffer. The external process writes to the frame buffer and signals a condition variable when the buffer is full. The video driver waits on the condition variable to synchronize the "frame rate" of the external application.

### CODE REVIEW

File: examples/SharedMemoryVideoWriter/main.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include <pangolin/utils/condition_variable.h>
+#include <pangolin/utils/shared_memory_buffer.h>
+#include <pangolin/utils/timer.h>
+
+#include <cmath>
+#include <cstdint>
+
+// This sample acts as a soft camera. It writes a pattern of GRAY8 pixels to the
+// shared memory space. It can be seen in Pangolin's SimpleVideo sample using
+// the shmem://example?size=640x480 video URI.
```
Comment (stevenlovegrove) [inline examples/SharedMemoryVideoWriter/main.cpp]: Pangolin Video URI's are of the form scheme:[option1=value1,option2=value2,...]//device_name

File: include/pangolin/utils/condition_variable.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <pangolin/utils/timer.h>
+
+#include <memory>
```
Comment (stevenlovegrove) [inline include/pangolin/utils/condition_variable.h]: Pangolin uses Boost if C++11 isn't available. We can't use unique_ptr, but shared_ptr is available via a compatibility header (https://github.com/stevenlovegrove/Pangolin/blob/master/include/pangolin/compat/memory.h)

File: src/utils/semaphore.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
+#include <pangolin/utils/semaphore.h>
+
+#include <fcntl.h>
+#include <semaphore.h>
+#include <sys/stat.h>
```
Comment (stevenlovegrove) [inline src/utils/semaphore.cpp]: Looks like this isn't going to build on Windows. Not sure about OSX. We will need to conditionally add this stuff depending on platform.

**Comment by stevenlovegrove** [comment]
This looks good Shaheen. We just need to avoid C++11 and add some platform config. I don't mind doing that myself so that you can get on and start using this.

**Comment by stevenlovegrove** [comment]
I just tested it on OSX and it works great!

**Comment by visigoth** [comment]
created a new development branch for myself, so closing this one and opening a new PR

**Pull request closed without merging**